technical-docs/2021-05-remote-restore-redesign.md

```diff
@@ -0,0 +1,110 @@
+# Medusa Operator - Restore redesign document
+
+## Background
+
+Medusa operator doesn’t yet support remote restores: restoring a backup to another cluster.
+As we implement this feature, we would like to have it support multiple use cases that operators could expect to have working:
+
+- Restoring a medusa-operator based backup to another Cassandra install in the same k8s cluster
+![](img/medusa_k8s-new-restore-design-1.jpg)
+
+- Restoring a medusa-operator based backup to another Cassandra install in a different k8s cluster
+![](img/medusa_k8s-new-restore-design-2.jpg)
+
+- Restoring a Medusa backup (non k8s) to a k8s Cassandra install
+![](img/medusa_k8s-new-restore-design-3.jpg)
+
+
+
+
+## Sync CassandraBackup resources with storage
+The target solution will need to be able to sync the local CassandraBackup resources with the configured storage backend.
+The sync should allow:
+
+- Creating local missing resources from existing backups
+- Deleting local resources for missing backups
+- Deleting remote backups if a local resource was marked as deleted
+
+A single control loop should be able to handle all 3 operations.
+The sync should be required explicitly, meaning that a new CRD should be created to request synchronization to the medusa-operator, and that CRD should contain a prefix so that a cluster can list backups stored for another cluster in a multi-tenant storage bucket.
+
+The CassandraBackup resource should be updated, adding a delete field which would allow marking backups for deletion.
```

Comment:
The way to mark an object for deletion is with finalizers. See this [post](https://kubernetes.io/blog/2021/05/14/using-finalizers-to-control-deletion/) for background.

Here is brief summary of how it works. When a CassandraBackup is created the backup controller adds a finalizer to the object. Kubernetes will not delete an object that has finalizers. When the CassandraBackup is deleted, Kubernetes sets a deletion timestamp on the object and reconciliation is triggered. The backup controller sees that the deletion timestamp is set and then proceeds with whatever clean up after is needed. When all of the clean up is done, the finalizer is removed. At this point Kubernetes can continue with the actual deletion process.

Reply:
Sounds good, I'll update the doc appropriately to use finalizers instead.

Reply:
I've updated the doc to use finalizers. Let me know how it looks.

Reply:
Looks good!